ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Iterator.hs

```diff
@@ -0,0 +1,301 @@
+import           Ouroboros.Storage.EpochInfo (fixedSizeEpochInfo)
+import qualified Ouroboros.Storage.ImmutableDB as ImmDB
+import qualified Ouroboros.Storage.Util.ErrorHandling as EH
+import qualified Ouroboros.Storage.VolatileDB as VolDB
+
+
+import           Test.Ouroboros.Storage.ChainDB.TestBlock
+import qualified Test.Ouroboros.Storage.ImmutableDB.Mock as ImmDB (openDBMock)
+import qualified Test.Ouroboros.Storage.VolatileDB.Mock as VolDB (openDBMock)
+
+{-------------------------------------------------------------------------------
+  Top-level tests
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "Iterator"
+    [ testProperty "#773 bug in example 1"  prop_773_bug
+    , testProperty "#773 correct example 2" prop_773_working
+    ]
+
+-- These tests focus on the implementation of the ChainDB iterators, which are
+-- used to stream blocks from the ChainDB. A few things make this code
+-- complex:
+--
+-- * We need to be able to stream from both the ImmutableDB and the
+--   VolatileDB.
+-- * While streaming, blocks might be copied from the VolatileDB to the
+--   ImmutableDB.
+-- * While streaming, blocks might be garbage-collected from the VolatileDB.
+--   These blocks might have been copied to the ImmutableDB or not.
+--
+-- The copying and garbage collection will happen in the background,
+-- /concurrently/ with the streaming, so we have to be careful about race
+-- conditions. For these reasons, we provide separate tests for the ChainDb
+-- iterators.
+--
+-- To avoid the complexity of a whole ChainDB and to have explicit control of
+-- the copying and garbage collection, we set up a mock 'IteratorEnv' record
+-- containing (amongst others) a mock ImmutableDB and a mock VolatileDB that
+-- can be manipulated directly, instead of relying on the background threads
+-- to manipulate them for us.
+
+-- TODO:
+-- - Write a generator for TestSetup and a model implementation (reuse
+--   ChainDB.Model) to turn this into a property test.
+-- - Instead of simply reading all blocks, use:
+--   > data Action = IterNext .. | CopyToImmDB .. | GCFromVolDB ..
+--   And write a generator for it.
+-- - Run multiple @Action@s in parallel
+
+{-------------------------------------------------------------------------------
+  Test cases
+-------------------------------------------------------------------------------}
+
+-- All blocks on the same chain
+a, b, c, d, e :: TestBlock
+a = mkBlk [0]
+b = mkBlk [0,0]
+c = mkBlk [0,0,0]
+d = mkBlk [0,0,0,0]
+e = mkBlk [0,0,0,0,0]
+
+-- | Requested stream = A -> C
+
+--              ImmDB              VolDB
+-- Hash    A -> B -> C -> D        C, D
+--
+-- Bug: we find a partial path [B]->C in the VolDB. Now the ForkTooOld
+-- condition is triggered because the tip of the ImmDB is not B but D.
+--
+-- For more details, see:
+-- https://github.com/input-output-hk/ouroboros-network/pull/773#issuecomment-513128004
+prop_773_bug :: Property
```

Comment:
This is the bug @nfrisby found in https://github.com/input-output-hk/ouroboros-network/pull/773#issuecomment-512923081 This test is failing at the moment.